packages/db/src/scheduler.ts

```diff
@@ -0,0 +1,204 @@
+        completed: new Set(),
+      }
+      this.contexts.set(contextId, context)
+    }
+    return context
+  }
+
+  /**
+   * Schedule work. Without a context id, executes immediately.
+   * Otherwise queues the job to be flushed once dependencies are satisfied.
+   */
+  schedule<TState>({
+    contextId,
+    jobId,
+    dependencies,
+    createEntry,
+    updateEntry,
+  }: ScheduleOptions<TState>): void {
+    if (typeof contextId === `undefined`) {
+      const entry = createEntry()
+      updateEntry?.(entry)
+      entry.run()
+      return
+    }
+
+    const context = this.getOrCreateContext(contextId)
+
+    let entry = context.entries.get(jobId) as SchedulerEntry<TState> | undefined
+    if (!entry) {
+      entry = createEntry()
+      context.entries.set(jobId, entry)
+      context.queue.push(jobId)
+    }
+
+    updateEntry?.(entry)
+
+    if (dependencies) {
+      const depSet = new Set<unknown>()
+      for (const dep of dependencies) {
+        if (dep !== jobId) {
+          depSet.add(dep)
+        }
+      }
+      context.dependencies.set(jobId, depSet)
+    } else if (!context.dependencies.has(jobId)) {
+      context.dependencies.set(jobId, new Set())
+    }
+
+    context.completed.delete(jobId)
+  }
+
+  /**
+   * Flush all queued work for a context. Jobs with unmet dependencies are retried.
+   * Throws if a pass completes without running any job (dependency cycle).
+   */
+  flush(contextId: SchedulerContextId): void {
+    const context = this.contexts.get(contextId)
+    if (!context) return
+
+    const { queue, entries, dependencies, completed } = context
+
+    while (queue.length > 0) {
+      let ranThisPass = false
+      const jobsThisPass = queue.length
+
+      for (let i = 0; i < jobsThisPass; i++) {
+        const jobId = queue.shift()!
+        const entry = entries.get(jobId)
+        if (!entry) {
+          dependencies.delete(jobId)
+          completed.delete(jobId)
+          continue
+        }
+
+        const deps = dependencies.get(jobId)
+        const ready =
+          !deps ||
+          deps.size === 0 ||
+          [...deps].every((dep) => dep === jobId || completed.has(dep))
+
+        if (ready) {
+          entries.delete(jobId)
+          dependencies.delete(jobId)
+          // Run the job. If it throws, we don't mark it complete, allowing the
+          // error to propagate while maintaining scheduler state consistency.
+          entry.run()
+          completed.add(jobId)
+          ranThisPass = true
+        } else {
+          queue.push(jobId)
+        }
+      }
+
+      if (!ranThisPass) {
+        throw new Error(
+          `Scheduler detected unresolved dependencies for context ${String(
+            contextId
+          )}.`
+        )
+      }
+    }
+
+    this.contexts.delete(contextId)
+  }
+
+  /**
+   * Flush all contexts with pending work. Useful during tear-down.
+   */
+  flushAll(): void {
+    for (const contextId of Array.from(this.contexts.keys())) {
```

Comment:
Given that `flush` is `O(n^2)` and this is calling `flush` for each `contextId` it becomes even more expensive.